module/zcommon/zfs_fletcher_aarch64_neon.c

```diff
@@ -69,133 +76,21 @@ fletcher_4_aarch64_neon_fini(fletcher_4_ctx_t *ctx, zio_cksum_t *zcp)
 	    8 * ctx->aarch64_neon[3].v[1] - 8 * ctx->aarch64_neon[2].v[1] +
 	    ctx->aarch64_neon[1].v[1];
 	ZIO_SET_CHECKSUM(zcp, A, B, C, D);
+	kfpu_end();
 }
 
-#define	NEON_INIT_LOOP()			\
-	asm("eor %[ZERO].16b,%[ZERO].16b,%[ZERO].16b\n"	\
-	"ld1 { %[ACC0].4s }, %[CTX0]\n"		\
-	"ld1 { %[ACC1].4s }, %[CTX1]\n"		\
-	"ld1 { %[ACC2].4s }, %[CTX2]\n"		\
-	"ld1 { %[ACC3].4s }, %[CTX3]\n"		\
-	: [ZERO] "=w" (ZERO),			\
-	[ACC0] "=w" (ACC0), [ACC1] "=w" (ACC1),	\
-	[ACC2] "=w" (ACC2), [ACC3] "=w" (ACC3)	\
-	: [CTX0] "Q" (ctx->aarch64_neon[0]),	\
-	[CTX1] "Q" (ctx->aarch64_neon[1]),	\
-	[CTX2] "Q" (ctx->aarch64_neon[2]),	\
-	[CTX3] "Q" (ctx->aarch64_neon[3]))
-
-#define	NEON_DO_REVERSE "rev32 %[SRC].16b, %[SRC].16b\n"
-
-#define	NEON_DONT_REVERSE ""
-
-#define	NEON_MAIN_LOOP(REVERSE)				\
-	asm("ld1 { %[SRC].4s }, %[IP]\n"		\
-	REVERSE						\
-	"zip1 %[TMP1].4s, %[SRC].4s, %[ZERO].4s\n"	\
-	"zip2 %[TMP2].4s, %[SRC].4s, %[ZERO].4s\n"	\
-	"add %[ACC0].2d, %[ACC0].2d, %[TMP1].2d\n"	\
-	"add %[ACC1].2d, %[ACC1].2d, %[ACC0].2d\n"	\
-	"add %[ACC2].2d, %[ACC2].2d, %[ACC1].2d\n"	\
-	"add %[ACC3].2d, %[ACC3].2d, %[ACC2].2d\n"	\
-	"add %[ACC0].2d, %[ACC0].2d, %[TMP2].2d\n"	\
-	"add %[ACC1].2d, %[ACC1].2d, %[ACC0].2d\n"	\
-	"add %[ACC2].2d, %[ACC2].2d, %[ACC1].2d\n"	\
-	"add %[ACC3].2d, %[ACC3].2d, %[ACC2].2d\n"	\
-	: [SRC] "=&w" (SRC),				\
-	[TMP1] "=&w" (TMP1), [TMP2] "=&w" (TMP2),	\
-	[ACC0] "+w" (ACC0), [ACC1] "+w" (ACC1),		\
-	[ACC2] "+w" (ACC2), [ACC3] "+w" (ACC3)		\
-	: [ZERO] "w" (ZERO), [IP] "Q" (*ip))
-
-#define	NEON_FINI_LOOP()			\
-	asm("st1 { %[ACC0].4s },%[DST0]\n"	\
-	"st1 { %[ACC1].4s },%[DST1]\n"		\
-	"st1 { %[ACC2].4s },%[DST2]\n"		\
-	"st1 { %[ACC3].4s },%[DST3]\n"		\
-	: [DST0] "=Q" (ctx->aarch64_neon[0]),	\
-	[DST1] "=Q" (ctx->aarch64_neon[1]),	\
-	[DST2] "=Q" (ctx->aarch64_neon[2]),	\
-	[DST3] "=Q" (ctx->aarch64_neon[3])	\
-	: [ACC0] "w" (ACC0), [ACC1] "w" (ACC1),	\
-	[ACC2] "w" (ACC2), [ACC3] "w" (ACC3))
-
 static void
 fletcher_4_aarch64_neon_native(fletcher_4_ctx_t *ctx,
```

Comment:
Since `kfpu_begin()`/`kfpu_end()` have been moved into the init and fini functions, we could stop writing a `fletcher_4_aarch64_neon_native()` shim that calls the real `fletcher_4_aarch64_neon_native_impl()` assembly function in C and instead just name the assembly function `fletcher_4_aarch64_neon_native()` and let `fletcher_4_aarch64_neon_ops` point to it. The same goes for the byteswap version.

My apologies for not clarifying this sooner. I did not realize until after I wrote the patch that moved `kfpu_begin()`/`kfpu_end()` into the ini and fini functions. After I wrote it, I forgot to post a revision to what I had said here.